### PR TITLE
Implement opening cloud scores (first iteration)

### PR DESCRIPTION
**Pull request by cbjeukendrup**
Resolves: https://github.com/musescore/MuseScore/issues/16808
Closes: https://github.com/musescore/MuseScore/issues/17672

Builds on: #17708 

Terminology: the "new" cloud scores folder is in the hidden user data folders, while the "old" cloud scores folder is a folder called "Cloud Scores" in the user's documents folder. We should not tempt the user so much to shoot themselves in the foot, as we do by storing the cloud scores folder in a user-visible location.

"New" cloud scores are stored under their scoreId, rather than their name, to prevent name conflicts (MuseScore.com supports multiple scores having the same name).
When the user tries to save a newly created score to the cloud, this score is first saved under a temporary name in the cloud scores folder, and once it is successfully uploaded and the scoreId becomes known, it's moved to its definitive name, namely its scoreId.

Known problems:
- [x] See also known problems of #17708 
- [x] Error handling not really implemented yet
- [ ] ~Cloud score is ALWAYS downloaded; no check for local version yet. No client-side open-time conflict handling. So, if you save a cloud score but the changes fail to be uploaded, and you then open it again, the online version will be downloaded and the local version with the changes is overwritten.~ We decided that this will be delayed to the future.
- [ ] ~As a consequence, the status indicator button is always blue (rather than black or blue based on the current situation)~ We decided that this will be delayed to the future.
- [ ] There is no protection for what will happen if the user tries to open another score while a cloud score is already being downloaded
- [x] In the Recent Scores tab, for "new" cloud scores, the scoreId is displayed instead of the displayName. This is the reason that I have used the `RecentFile` typealias rather than `io::path_t` directly: `RecentFile` will become a struct that contains another member, namely the display name. This display name will be stored in the recent files JSON data stored in the configuration.
- [x] Do something nice about converting "old" cloud scores to "new" cloud scores?
- [ ] Most scores download so fast that the download progress bar does not even get the chance to appear. (Actually opening the score takes longer than downloading it, therefore the score does still not appear instantly.) So, this needs testing with big scores and a slow internet connection.
- [ ] Needs thorough testing

### CODE REVIEW

File: src/framework/global/progress.h
Comment (Eism) [inline src/framework/global/progress.h:40]: May be 'canceled'?
Because the action has already happened

**Comment by Eism** [comment]
It is better to move commits with fixing of warnings to a separate PR

File: src/project/internal/projectconfiguration.cpp
Comment (Eism) [inline src/project/internal/projectconfiguration.cpp:236]: is it a temporary solution?
If not, then please explain how it works

File: src/project/internal/projectconfiguration.cpp
Comment (Eism) [inline src/project/internal/projectconfiguration.cpp:272]: counter++

File: src/project/internal/projectconfiguration.cpp
Comment (cbjeukendrup) [reply src/project/internal/projectconfiguration.cpp:236]: It is temporary. As mentioned in the PR description, we should put cloud scores in a hidden location; that is the `newCloudProjectsPath()`. But we also want that existing cloud scores are still treated as cloud scores. Maybe we should add something about compatibility, to migrate "old" cloud scores to the "new" cloud folder. Or we should do this migration at the moment that the user saves such an old cloud score, and convert it into a new one at that moment.

File: src/framework/global/progress.h
Comment (cbjeukendrup) [reply src/framework/global/progress.h:42]: Not really... I've added a comment to the code to clarify.

**Comment by cbjeukendrup** [comment]
I'm very not proud of the last commit, but I'm really running a bit out of inspiration.

File: src/app/commandlineparser.cpp
Comment (RomanPudashkin) [inline src/app/commandlineparser.cpp:141]: What problem are you trying to solve with this option?

File: src/project/internal/projectactionscontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/projectactionscontroller.cpp:54]: What is the difference between file-open and file-open-projectfile?

File: src/cloud/musescorecom/musescorecomservice.cpp
Comment (RomanPudashkin) [inline src/cloud/musescorecom/musescorecomservice.cpp:336]: Please uncomment this code

File: src/project/internal/projectconfiguration.h
Comment (RomanPudashkin) [inline src/project/internal/projectconfiguration.h:76]: Why do we need two of them in the interface? It's better to keep only isCloudProject and hide all the details behind its implementation

File: src/project/internal/recentfilescontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/recentfilescontroller.cpp:97]: Why are these arguments of a different type?

File: src/app/commandlineparser.cpp
Comment (cbjeukendrup) [reply src/app/commandlineparser.cpp:141]: When loading a new instance of MuseScore because the user opens another file while there is already a file open, a special splash screen is displayed, containing the filename of the score being opened. 
For cloud scores, the filename is the score id (this is required to fix multiple problems, namely that multiple MuseScore.com scores can have the same name and that those names can contain characters that are not safe for a filename). 
But in the UI we want to see the real name for these cloud scores. 
At the time that the splash screen is created, we don't yet have initialised enough stuff to download the name information from MuseScore.com, so we need to pass it via a command line option from the existing instance of MuseScore. At least, that was the best I could think of.

File: src/project/internal/recentfilescontroller.cpp
Comment (cbjeukendrup) [reply src/project/internal/recentfilescontroller.cpp:97]: Only the file path is really a key, and the displayNameOverride is some additional data. So, to look up the existing file in the list, we should only compare the path.

File: src/project/internal/projectconfiguration.h
Comment (cbjeukendrup) [reply src/project/internal/projectconfiguration.h]: I cleaned this up a little bit, and ensured that it also works well when opening "legacy" cloud scores (from the old cloud scores folder from 4.0)

File: src/project/internal/recentprojectscontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/recentprojectscontroller.cpp:1]: It looks like this file can be deleted

File: src/appshell/internal/startupscenario.cpp
Comment (RomanPudashkin) [inline src/appshell/internal/startupscenario.cpp:68]: path -> file

File: src/appshell/internal/istartupscenario.h
Comment (RomanPudashkin) [inline src/appshell/internal/istartupscenario.h:41]: Could you please change it to` const project::ProjectFile&`?

File: src/appshell/view/appmenumodel.cpp
Comment (RomanPudashkin) [inline src/appshell/view/appmenumodel.cpp:408]: Please add a comment to explain what "true" means here

File: src/app/commandlineparser.cpp
Comment (RomanPudashkin) [reply src/app/commandlineparser.cpp:141]: You should write some description for this option. Also, you can hide it from -h (to make it really an internal option)

File: src/project/internal/recentfilescontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/recentfilescontroller.cpp:226]: You should use our implementation of JsonArray/JsonObject (they are in json.h)

File: src/project/internal/recentfilescontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/recentfilescontroller.cpp:248]: file -> filePath

File: src/project/internal/projectactionscontroller.h
Comment (RomanPudashkin) [inline src/project/internal/projectactionscontroller.h:86]: const ProjectBeingDownloaded&

File: src/project/internal/recentfilescontroller.cpp
Comment (RomanPudashkin) [inline src/project/internal/recentfilescontroller.cpp:160]: You should use emplace_back

**Comment by bkunda** [comment]
Tested primarily on macOS, and a bit in Windows:

All seems to be fine from my POV.

Waiting on final approval from @DmitryArefiev and @jessjwilliamson before we can merge. 

Thanks!

**Comment by DmitryArefiev** [comment]
Tested on **Windows10, LinuxUbuntu22.04**, all fine on my side

**Comment by cbjeukendrup** [comment]
Code review fixes pushed

**Review by RomanPudashkin — APPROVED** [review]

**Comment by jessjwilliamson** [comment]
Yes all good from me :)

**Pull request merged**
